### ISSUE TITLE
[Dask] Prefect is unable to handle rescheduling events or worker retirement 

### ISSUE THREAD

**Issue by BitTheByte** [PrefectHQ/prefect#9014]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

I think the title pretty much explains everything, but to be more clear. prefect is reporting `prefect.exceptions.CrashedRun` Execution was cancelled by the runtime environment." as a failure **even if retires are configured** when a dask decides to retire workers and move the execution somewhere else.

Related to:
https://github.com/PrefectHQ/prefect/issues/9012
https://github.com/PrefectHQ/prefect/issues/8821

### Reproduction

```python3
import asyncio

from prefect import flow, task
from prefect_dask.task_runners import DaskTaskRunner
from dask.distributed import Scheduler
from time import sleep


class PoCDaskTaskRunner(DaskTaskRunner):
    def get_scheduler(self) -> "Scheduler":
        return self._cluster.scheduler

    async def submit(self, *args, **kwargs):

        await super().submit(*args, **kwargs)

        await asyncio.sleep(1)  # allow some time for submit to finish

        scheduler = self.get_scheduler()
        worker = list(scheduler.workers.keys())[0]

        # to simulate worker being retired, and its future being moved to another one
        await scheduler.retire_workers(workers=[worker])


task_runner = PoCDaskTaskRunner(
    cluster_class="dask.distributed.LocalCluster",
    cluster_kwargs={
        "processes": False,
        "n_workers": 2,
        "threads_per_worker": 1
    }
)


@task(retries=100, retry_delay_seconds=1)  # this take should be able to run 100 times!
def dummy_task():
    print("DUMMY TASK IS RUNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNIIINNNNGGG!!!")
    sleep(5)


@flow(task_runner=task_runner, persist_result=True)
def test_flow():
    dummy_task.submit()


if __name__ == "__main__":
    test_flow()
```

### Error

```python3
16:31:33.081 | INFO    | prefect.engine - Created flow run 'laughing-boa' for flow 'test-flow'
16:31:33.082 | INFO    | prefect.task_runner.pocdask - Creating a new Dask cluster with `distributed.deploy.local.LocalCluster`
16:31:33.565 | INFO    | distributed.scheduler - State start
16:31:33.567 | INFO    | distributed.scheduler -   Scheduler at: inproc://192.168.158.81/3785/1
16:31:33.568 | INFO    | distributed.scheduler -   dashboard at:  http://192.168.158.81:8787/status
16:31:33.574 | INFO    | distributed.worker -       Start worker at: inproc://192.168.158.81/3785/4
16:31:33.574 | INFO    | distributed.worker -          Listening to:       inproc192.168.158.81
16:31:33.574 | INFO    | distributed.worker -           Worker name:                          0
16:31:33.575 | INFO    | distributed.worker -          dashboard at:       192.168.158.81:41683
16:31:33.575 | INFO    | distributed.worker - Waiting to connect to: inproc://192.168.158.81/3785/1
16:31:33.575 | INFO    | distributed.worker - -------------------------------------------------
16:31:33.576 | INFO    | distributed.worker -               Threads:                          1
16:31:33.576 | INFO    | distributed.worker -                Memory:                   5.85 GiB
16:31:33.576 | INFO    | distributed.worker -       Local Directory: /tmp/dask-worker-space/worker-x6mt81dt
16:31:33.576 | INFO    | distributed.worker - -------------------------------------------------
16:31:33.577 | INFO    | distributed.worker -       Start worker at: inproc://192.168.158.81/3785/5
16:31:33.577 | INFO    | distributed.worker -          Listening to:       inproc192.168.158.81
16:31:33.577 | INFO    | distributed.worker -           Worker name:                          1
16:31:33.578 | INFO    | distributed.worker -          dashboard at:       192.168.158.81:38127
16:31:33.578 | INFO    | distributed.worker - Waiting to connect to: inproc://192.168.158.81/3785/1
16:31:33.578 | INFO    | distributed.worker - -------------------------------------------------
16:31:33.578 | INFO    | distributed.worker -               Threads:                          1
16:31:33.578 | INFO    | distributed.worker -                Memory:                   5.85 GiB
16:31:33.579 | INFO    | distributed.worker -       Local Directory: /tmp/dask-worker-space/worker-_v3cw1je
16:31:33.579 | INFO    | distributed.worker - -------------------------------------------------
16:31:33.602 | INFO    | distributed.scheduler - Register worker <WorkerState 'inproc://192.168.158.81/3785/4', name: 0, status: init, memory: 0, processing: 0>
16:31:33.613 | INFO    | distributed.scheduler - Starting worker compute stream, inproc://192.168.158.81/3785/4
16:31:33.613 | INFO    | distributed.core - Starting established connection to inproc://192.168.158.81/3785/6
16:31:33.614 | INFO    | distributed.scheduler - Register worker <WorkerState 'inproc://192.168.158.81/3785/5', name: 1, status: init, memory: 0, processing: 0>
16:31:33.624 | INFO    | distributed.scheduler - Starting worker compute stream, inproc://192.168.158.81/3785/5
16:31:33.624 | INFO    | distributed.core - Starting established connection to inproc://192.168.158.81/3785/7
16:31:33.625 | INFO    | distributed.worker -         Registered to: inproc://192.168.158.81/3785/1
16:31:33.625 | INFO    | distributed.worker - -------------------------------------------------
16:31:33.625 | INFO    | distributed.worker -         Registered to: inproc://192.168.158.81/3785/1
16:31:33.625 | INFO    | distributed.worker - -------------------------------------------------
16:31:33.626 | INFO    | distributed.core - Starting established connection to inproc://192.168.158.81/3785/1
16:31:33.626 | INFO    | distributed.core - Starting established connection to inproc://192.168.158.81/3785/1
16:31:33.637 | INFO    | distributed.scheduler - Receive client connection: Client-bc37b622-cfd0-11ed-8ec9-00155d56716e
16:31:33.646 | INFO    | distributed.core - Starting established connection to inproc://192.168.158.81/3785/8
16:31:33.647 | INFO    | prefect.task_runner.pocdask - The Dask dashboard is available at http://192.168.158.81:8787/status
16:31:33.778 | INFO    | Flow run 'laughing-boa' - Created task run 'dummy_task-0' for task 'dummy_task'
DUMMY TASK IS RUNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNIIINNNNGGG!!!
16:31:34.989 | INFO    | distributed.scheduler - Retiring worker inproc://192.168.158.81/3785/4
16:31:34.990 | INFO    | distributed.active_memory_manager - Retiring worker inproc://192.168.158.81/3785/4; no unique keys need to be moved away.
16:31:34.990 | INFO    | distributed.scheduler - Remove worker <WorkerState 'inproc://192.168.158.81/3785/4', name: 0, status: closing_gracefully, memory: 0, processing: 1>
16:31:34.991 | INFO    | distributed.core - Removing comms to inproc://192.168.158.81/3785/4
16:31:34.991 | INFO    | distributed.scheduler - Retired worker inproc://192.168.158.81/3785/4
16:31:34.991 | INFO    | Flow run 'laughing-boa' - Submitted task run 'dummy_task-0' for execution.
16:31:34.992 | INFO    | distributed.core - Removing comms to inproc://192.168.158.81/3785/4
DUMMY TASK IS RUNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNIIINNNNGGG!!!
2023-03-31 16:31:35,127 - distributed.scheduler - WARNING - Received heartbeat from unregistered worker 'inproc://192.168.158.81/3785/4'.
16:31:35.127 | WARNING | distributed.scheduler - Received heartbeat from unregistered worker 'inproc://192.168.158.81/3785/4'.
2023-03-31 16:31:35,128 - distributed.worker - ERROR - Scheduler was unaware of this worker 'inproc://192.168.158.81/3785/4'. Shutting down.
16:31:35.128 | ERROR   | distributed.worker - Scheduler was unaware of this worker 'inproc://192.168.158.81/3785/4'. Shutting down.
16:31:35.128 | INFO    | distributed.worker - Stopping worker at inproc://192.168.158.81/3785/4. Reason: worker-close
16:31:35.129 | ERROR   | Task run 'dummy_task-0' - Crash detected! Execution was cancelled by the runtime environment.
16:31:35.151 | INFO    | distributed.worker - Async task 'dummy_task-0-89c05f5fb2ff428abc013499a52caff3-1' cancelled during worker close; rescheduling.
16:31:35.152 | INFO    | distributed.core - Received 'close-stream' from inproc://192.168.158.81/3785/6; closing.
16:31:35.153 | INFO    | distributed.core - Connection to inproc://192.168.158.81/3785/1 has been closed.
16:31:40.058 | INFO    | Task run 'dummy_task-0' - Task run '89c05f5f-b2ff-428a-bc01-3499a52caff3' already finished.
16:31:40.097 | INFO    | distributed.scheduler - Remove client Client-bc37b622-cfd0-11ed-8ec9-00155d56716e
16:31:40.097 | INFO    | distributed.core - Received 'close-stream' from inproc://192.168.158.81/3785/8; closing.
16:31:40.097 | INFO    | distributed.scheduler - Remove client Client-bc37b622-cfd0-11ed-8ec9-00155d56716e
16:31:40.098 | INFO    | distributed.scheduler - Close client connection: Client-bc37b622-cfd0-11ed-8ec9-00155d56716e
16:31:40.099 | INFO    | distributed.worker - Stopping worker at inproc://192.168.158.81/3785/5. Reason: worker-close
16:31:40.099 | INFO    | distributed.core - Received 'close-stream' from inproc://192.168.158.81/3785/7; closing.
16:31:40.100 | INFO    | distributed.scheduler - Remove worker <WorkerState 'inproc://192.168.158.81/3785/5', name: 1, status: closing, memory: 0, processing: 0>
16:31:40.100 | INFO    | distributed.core - Removing comms to inproc://192.168.158.81/3785/5
16:31:40.100 | INFO    | distributed.scheduler - Lost all workers
16:31:40.100 | INFO    | distributed.core - Connection to inproc://192.168.158.81/3785/1 has been closed.
16:31:40.101 | INFO    | distributed.scheduler - Scheduler closing...
16:31:40.102 | INFO    | distributed.scheduler - Scheduler closing all comms
16:31:40.102 | ERROR   | Flow run 'laughing-boa' - Finished in state Failed('1/1 states failed.')
---------------------------------------------------------------------------
CrashedRun                                Traceback (most recent call last)
File /mnt/e/Dev/emprax/workflow/event_loop_poc.py:50
     46     dummy_task.submit()
     49 if __name__ == "__main__":
---> 50     test_flow()

File ~/.cache/pypoetry/virtualenvs/emprax-3XS7TJL0-py3.10/lib/python3.10/site-packages/prefect/flows.py:468, in Flow.__call__(self, return_state, wait_for, *args, **kwargs)
    464 parameters = get_call_parameters(self.fn, args, kwargs)
    466 return_type = "state" if return_state else "result"
--> 468 return enter_flow_run_engine_from_flow_call(
    469     self,
    470     parameters,
    471     wait_for=wait_for,
    472     return_type=return_type,
    473 )

File ~/.cache/pypoetry/virtualenvs/emprax-3XS7TJL0-py3.10/lib/python3.10/site-packages/prefect/engine.py:182, in enter_flow_run_engine_from_flow_call(flow, parameters, wait_for, return_type)
    177     retval = from_async.wait_for_call_in_loop_thread(
    178         begin_run, done_callbacks=done_callbacks
    179     )
    181 else:
--> 182     retval = from_sync.wait_for_call_in_loop_thread(
    183         begin_run, done_callbacks=done_callbacks
    184     )
    186 return retval

File ~/.cache/pypoetry/virtualenvs/emprax-3XS7TJL0-py3.10/lib/python3.10/site-packages/prefect/_internal/concurrency/api.py:137, in from_sync.wait_for_call_in_loop_thread(call, timeout, done_callbacks)     
    135     waiter.add_done_callback(callback)
    136 waiter.wait()
--> 137 return call.result()

File ~/.cache/pypoetry/virtualenvs/emprax-3XS7TJL0-py3.10/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py:173, in Call.result(self, timeout)
    167 def result(self, timeout: Optional[float] = None) -> T:
    168     """
    169     Wait for the result of the call.
    170
    171     Not safe for use from asynchronous contexts.
    172     """
--> 173     return self.future.result(timeout=timeout)

File /usr/lib/python3.10/concurrent/futures/_base.py:451, in Future.result(self, timeout)
    449     raise CancelledError()
    450 elif self._state == FINISHED:
--> 451     return self.__get_result()
    453 self._condition.wait(timeout)
    455 if self._state in [CANCELLED, CANCELLED_AND_NOTIFIED]:

File /usr/lib/python3.10/concurrent/futures/_base.py:403, in Future.__get_result(self)
    401 if self._exception:
    402     try:
--> 403         raise self._exception
    404     finally:
    405         # Break a reference cycle with the exception in self._exception
    406         self = None

File ~/.cache/pypoetry/virtualenvs/emprax-3XS7TJL0-py3.10/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py:218, in Call._run_async(***failed resolving arguments***)
    216                 logger.debug("%r using async cancel scope %r", self, ctx)
    217                 ctx.chain(self.cancel_context, bidirectional=True)
--> 218                 result = await coro
    219 except BaseException as exc:
    220     logger.debug("Encountered exception in async call %r", self, exc_info=True)

File ~/.cache/pypoetry/virtualenvs/emprax-3XS7TJL0-py3.10/lib/python3.10/site-packages/prefect/client/utilities.py:40, in inject_client.<locals>.with_injected_client(*args, **kwargs)
    257     return state
    258 elif return_type == "result":
--> 259     return await state.result(fetch=True)
    260 else:
    261     raise ValueError(f"Invalid return type for flow engine {return_type!r}.")

File ~/.cache/pypoetry/virtualenvs/emprax-3XS7TJL0-py3.10/lib/python3.10/site-packages/prefect/states.py:91, in _get_state_result(state, raise_on_failure)
     86     raise PausedRun("Run paused.")
     88 if raise_on_failure and (
     89     state.is_crashed() or state.is_failed() or state.is_cancelled()
     90 ):
---> 91     raise await get_state_exception(state)
     93 if isinstance(state.data, DataDocument):
     94     result = result_from_state_with_data_document(
     95         state, raise_on_failure=raise_on_failure
     96     )

CrashedRun: Execution was cancelled by the runtime environment.
```

### Versions

```Text
Version:             2.9.0
API version:         0.8.4                    
Python version:      3.10.8                   
Git commit:          69f57bd5                 
Built:               Thu, Mar 30, 2023 1:08 PM
OS/Arch:             linux/x86_64             
Profile:             default                  
Server type:         ephemeral                
Server:                                       
  Database:          sqlite                   
  SQLite version:    3.40.0
```

**Comment by BitTheByte** [comment]
So what being happening here

1) Task(A) submitted on worker X 
2) Task(A) starts executing on worker X
3) Scheduler decided to retire worker X and move its workload to worker Y with Task(A) mid-running
4) Prefect decides that "Execution was cancelled by the runtime environment" and marks the task as failed
5) Task(A) starts executing on worker Y
6) Prefect allows the task to start again on worker Y and then reports the old task run state ???

**Comment by BitTheByte** [comment]
More debugging revealed that `_get_state_result` receives both the "Completed" and "Crashed" states but it choose to the "Crashed" saying that this take already finished before

**Comment by taylor-curran** [comment]
Many users use spot instances which will end up killing Dask Workers fairly often. Their workloads get migrated and retried on other machines, but Prefect is sometimes unable to handle running the task twice. The first time it is ran, prefect may have moved that task into a  Crashed state while on the worker node, but before it returns back to the Flow node (which is where tasks get submitted/awaited by prefect), dask can crash and reschedule the task because the task wasn't registered as finished by the dask scheduler.
For example, a given task run -- the second run will print "Task run '<task run ID>' already finished.", but really the users need Dask to be able to rerun it and move it back to a Running state.

**Comment by madkinsz** [comment]
This will be closed by #9152

**Comment by BitTheByte** [comment]
Resolved by `2.10.6`
